File: scripts/dev.mjs
#!/usr/bin/env node
import { spawn } from 'node:child_process';
import { existsSync } from 'node:fs';

/**
 * `npm run dev` — every service in one terminal, with its own database role so
 * row-level security applies exactly as it does in production.
 *
 * Under compose every service is handed its environment by the orchestrator.
 * Here nothing does, so the parent loads `.env` once and the children inherit
 * it — otherwise every service boots and immediately dies on a missing
 * DATABASE_URL, which is a confusing way to learn that a file was not read.
 */
if (existsSync('.env')) {
  process.loadEnvFile('.env');
} else {
  console.error('no .env found — copy .env.example and fill it in');
  process.exit(1);
}
const SERVICES = [
  ['gateway',    'services/gateway/src/index.ts',    'loop_gateway',    '3000'],
  ['connector',  'services/connector/src/index.ts',  'loop_connector',  '9105'],
  ['classifier', 'services/classifier/src/index.ts', 'loop_classifier', '9101'],
  ['extractor',  'services/extractor/src/index.ts',  'loop_extractor',  '9102'],
  ['resolver',   'services/resolver/src/index.ts',   'loop_resolver',   '9103'],
  ['pipeline',   'services/pipeline/src/index.ts',   'loop_pipeline',   '9104'],
  ['nudge',      'services/nudge/src/index.ts',      'loop_nudge',      '9106'],
  ['notifier',   'services/notifier/src/index.ts',   'loop_notifier',   '9107'],
];

const WIDTH = Math.max(...SERVICES.map(([n]) => n.length));
const children = [];

for (const [name, file, role, port] of SERVICES) {
  const child = spawn('node', ['--import', 'tsx', file], {
    env: { ...process.env, DB_ROLE: role, HEALTH_PORT: port },
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  const prefix = `${name.padEnd(WIDTH)} │ `;
  const pipe = (stream, out) =>
    stream.on('data', (d) =>
      out.write(String(d).split('\n').filter(Boolean).map((l) => prefix + l).join('\n') + '\n'),
    );
  pipe(child.stdout, process.stdout);
  pipe(child.stderr, process.stderr);
  child.on('exit', (code) => console.error(`${prefix}exited with ${code}`));
  children.push(child);
}

const stop = () => { for (const c of children) c.kill('SIGTERM'); };
process.on('SIGINT', () => { stop(); process.exit(0); });
process.on('SIGTERM', () => { stop(); process.exit(0); });

File: backend/scripts/dev.py
"""Every service in one terminal, each with its own database role.

    uv run --extra api --extra db --extra connector --extra push --extra ladder \
        python scripts/dev.py

Under compose every service is handed its environment by the orchestrator. Here
nothing does, so this loads `.env` once and the children inherit it — otherwise
every service boots and immediately dies on a missing `DATABASE_URL`, which is a
confusing way to learn that a file was not read.

The roles are the point. Each service connects as its own, so row-level security
and the per-service grants apply exactly as they do in production: run the
resolver and try to append an event and Postgres says no, here as there. Running
them all in one process would share one pool and one role, and every grant in
migration 003 would go back to being decorative.
"""

import argparse
import os
import signal
import subprocess
import sys
import threading
from contextlib import suppress
from pathlib import Path

# The API first, because it is the one you open. `python -m loop <name>` maps
# each of these to its own role in `loop/__main__.py`, so nothing is repeated.
SERVICES = (
    "api",
    "connector",
    "classifier",
    "extractor",
    "resolver",
    "pipeline",
    "nudge",
    "notifier",
)

WIDTH = max(len(name) for name in SERVICES)


def repo_root() -> Path:
    """The repository, where `.env` lives — one above `backend/`."""
    return Path(__file__).resolve().parents[2]


def load_env(path: Path) -> dict[str, str]:
    """`.env`, read the way compose reads it: `KEY=value`, no interpolation.

    Including the precedence. Compose resolves the shell environment *ahead* of
    the file, and seeding from `os.environ` and then writing over it does the
    opposite — so `DATABASE_URL=… python scripts/dev.py`, which is how the
    README says to point the fleet at the throwaway test database, silently ran
    against whatever is committed in `.env`.
    """
    if not path.is_file():
        raise SystemExit(f"no {path.name} found — copy .env.example and fill it in")
    env: dict[str, str] = {}
    for line in path.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#") or "=" not in stripped:
            continue
        key, _, value = stripped.partition("=")
        env[key.strip()] = _unquoted(value.strip())
    return env | os.environ


def _unquoted(value: str) -> str:
    """One matched pair of quotes, not every quote at either end."""
    for quote in ('"', "'"):
        if len(value) >= 2 and value.startswith(quote) and value.endswith(quote):
            return value[1:-1]
    return value


def pump(name: str, stream: object, out: object) -> None:
    prefix = f"{name:<{WIDTH}} │ "
    for line in stream:  # type: ignore[attr-defined]
        out.write(prefix + line)  # type: ignore[attr-defined]
        out.flush()  # type: ignore[attr-defined]


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--env-file", type=Path, default=repo_root() / ".env")
    parser.add_argument(
        "--only", nargs="*", choices=SERVICES, help="start a subset, for debugging one"
    )
    args = parser.parse_args()

    env = load_env(args.env_file)
    wanted = tuple(args.only) if args.only else SERVICES

    children: list[subprocess.Popen[str]] = []
    for name in wanted:
        child = subprocess.Popen(
            [sys.executable, "-m", "loop", name],
            env=env,
            cwd=str(Path(__file__).resolve().parents[1]),
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            bufsize=1,
        )
        children.append(child)
        threading.Thread(
            target=pump, args=(name, child.stdout, sys.stdout), daemon=True
        ).start()

    def stop(*_: object) -> None:
        for child in children:
            with suppress(ProcessLookupError):
                child.send_signal(signal.SIGTERM)

    signal.signal(signal.SIGINT, stop)
    signal.signal(signal.SIGTERM, stop)

    try:
        for child, name in zip(children, wanted, strict=True):
            code = child.wait()
            if code != 0:
                print(f"{name:<{WIDTH}} │ exited with {code}", file=sys.stderr)
    finally:
        stop()


if __name__ == "__main__":
    main()
